plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMetadata.java

```diff
@@ -303,13 +313,62 @@ private JdbcTableHandle flushAttributesAsQuery(ConnectorSession session, JdbcTab
         handle = new JdbcTableHandle(
                 handle.getRelationHandle(),
                 handle.getConstraint(),
+                handle.getSortOrder(),
                 OptionalLong.of(limit),
                 handle.getColumns(),
                 handle.getNextSyntheticColumnId());
 
         return Optional.of(new LimitApplicationResult<>(handle, jdbcClient.isLimitGuaranteed(session)));
     }
 
+    @Override
+    public Optional<TopNApplicationResult<ConnectorTableHandle>> applyTopN(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            long topNCount,
+            List<SortItem> sortItems,
+            Map<String, ColumnHandle> assignments)
+    {
+        if (!isTopNPushdownEnabled(session)) {
+            return Optional.empty();
+        }
+
+        verify(!sortItems.isEmpty(), "sortItems are empty");
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        List<SortItem> resultSortOrder = sortItems.stream()
+                .map(sortItem -> {
+                    verify(assignments.containsKey(sortItem.getName()), "assignments does not contain sortItem: %s", sortItem.getName());
+                    return new SortItem(((JdbcColumnHandle) assignments.get(sortItem.getName())).getColumnName(), sortItem.getSortOrder());
+                })
+                .collect(toImmutableList());
+
+        if (!jdbcClient.supportsTopN(session, handle, assignments, resultSortOrder)) {
```

Comment:
The `assignments` has been consumed (effecting in `resultSortOrder`), so no reason to pass them here.

Reply:
Yes, but we will lose information about the type that column holds (and won't be able to make decisions based on that)

Reply:
How would i get the column type base on the assignments map?

Reply:
As `Map<String, ColumnHandle>` has `JdbcColumnHandle` and `JdbcColumnHandle` contains a `Type`

Reply:
How do i find `JdbcColumnHandle` in that map for given `SortItem`?

Reply:
I've removed this field.